### ISSUE TITLE
Star Wars Battlefront: Elite Squadron can freeze PPSSPP at some point

### ISSUE THREAD

**Issue by Robsoie** [hrydgard/ppsspp#8076]
I have been playing Star Wars Battlefront: Elite Squadron, using the latest build for x86, that is v1.1.1-41-g9890460

The game play as good as usual in instant action, you can complete battles without a problem, galactic conquest seems correct too but in the campaign mode, after many missions without a single ppsspp bug (i'm amazed at how good ppsspp display that game with the improved resolution, played for hours without problem) but there's one in which you infiltrate an imperial destroyer, go to its ion canon and use it to destroy the defense on the ground, go back to your x-wing fighter and leave the destroyer with it with troopers are attacking you.

But with PPSSPP, as soon as your x-wing is going to leave the destroyer and the cutscene for it start, PPSSPP seems to go in a huge spam of these in the console 
W[FileSys] FileSystems\ISOFileSystem.cpp:571 Reading beyond end of file, clamping size 32768 to 2020
it's constantly repeating with several lines exactly similar added by seconds

before that line, you have sometime 
Atrac File-C
or
idle0
or
user_main
but after those the W[FileSys] thing is the exact same message

I can reproduce it everytime at that point of the campaign so i guess it's not a random bug, here is the save state if it can help 
https://www.dropbox.com/s/ghn8znsiaaalqa2/bf2efcrash.zip?dl=0

It's for the game in french language, i do not know if that save state can work on other variant
I added the picture in the zip that come with the save state file in case it's needed to load it

**Comment by daniel229** [comment]
Yes,it's freezing after x-wing fighter leave.But a gamesave is better to find out if it worked on the old build.

**Comment by daniel229** [comment]
Those logs seems to related,if it doesn't log those things,you can complete the mission.
```
W[FileSys] FileSystems\ISOFileSystem.cpp:571 Reading beyond end of file, clamping size 32768 to 2020
```
![01](https://cloud.githubusercontent.com/assets/3481559/10556556/9f82e368-74b8-11e5-851b-7baee61b2e72.png)

**Comment by unknownbrackets** [comment]
Does it log anything interesting looking before the first `Reading beyond end of file`?  Could the iso be a truncated dump?

-[Unknown]

**Comment by daniel229** [comment]
I don't see anything interesting before the first <code>Reading beyond end of file</code>,but there is a atrac3 error between them,and the background music stopped.
![02](https://cloud.githubusercontent.com/assets/3481559/10557060/27cb491c-74ce-11e5-9d25-fe4e7c9571cd.png)

**Comment by unknownbrackets** [comment]
That atrac error is very likely related.  Ultimately, it's probably causing it to die.

Very likely it's corrupt data.  What if you put this after the last `#include` in sceAtrac.cpp?

```c++
#undef DEBUG_LOG
#define DEBUG_LOG NOTICE_LOG
```

This will make it log sceAtrac more verbosely.  What does it say near the errors?

-[Unknown]

**Comment by daniel229** [comment]
something like this.
![03](https://cloud.githubusercontent.com/assets/3481559/10557172/c550340e-74d3-11e5-8f54-a74c73c2ecfa.png)

**Comment by unknownbrackets** [comment]
It's never actually calling sceAtracAddStreamData.  It might be an issue even with the remainingFrames value.  Anyway, pretty sure this is sceAtrac related.

-[Unknown]

**Comment by unknownbrackets** [comment]
Does the log look any different for this in the latest git build?  Or still the same error?

You can now use `#define HLE_LOG_FORCE` at the top of a file to force logging on for the entire file.

-[Unknown]

**Comment by daniel229** [comment]
Yes still the same.
![01](https://cloud.githubusercontent.com/assets/3481559/12046804/3b0096c4-aefb-11e5-8c43-970a8f564524.png)

full log rename jpg to 7z
![ppsspplog](https://cloud.githubusercontent.com/assets/3481559/12046808/557f7038-aefb-11e5-8624-ba623b11ec44.jpg)

**Comment by unknownbrackets** [comment]
Has this perhaps improved in the latest version, now that `sceAtracGetStreamDataInfo()` provides more reasonable values?

-[Unknown]

**Comment by daniel229** [comment]
The log
[PPSSPPWindows64v1.1.1-736-g14fb4f2.txt](https://github.com/hrydgard/ppsspp/files/93782/PPSSPPWindows64v1.1.1-736-g14fb4f2.txt)

**Comment by unknownbrackets** [comment]
Here's that trimmed down:

```
0=sceAtracSetDataAndGetID(08f62db0, 00001000): atrac3+ mono audio
00000000=sceAtracDecodeData(0, 08d4d084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f421d0], 09faa990[000002e8], 09faa994[0014e9f4])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d4f084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f424b8], 09faa990[000002e8], 09faa994[0014ecdc])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d51084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f427a0], 09faa990[000002e8], 09faa994[0014efc4])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d4d084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f41c00], 09faa990[000002e8], 09faa994[0014f2ac])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d4f084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f41ee8], 09faa990[000002e8], 09faa994[0014f594])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d51084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f421d0], 09faa990[000002e8], 09faa994[0014f87c])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d4d084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f424b8], 09faa990[000002e8], 09faa994[0014fb64])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d4f084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f427a0], 09faa990[000002e8], 09faa994[0014fe4c])
0=sceAtracAddStreamData(0, 000002e8)
00000000=sceAtracDecodeData(0, 08d51084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f41c00], 09faa990[000002e8], 09faa994[00150134])
0=sceAtracAddStreamData(0, 000002e8)
** Reading beyond end of file, clamping size 32768 to 3656
00000000=sceAtracDecodeData(0, 08d4d084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f41ee8], 09faa990[000002e8], 09faa994[0015041c])
0=sceAtracAddStreamData(0, 000002e8)
** Reading beyond end of file, clamping size 32768 to 2976
00000000=sceAtracDecodeData(0, 08d4f084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f421d0], 09faa990[000002e8], 09faa994[00150704])
0=sceAtracAddStreamData(0, 000002e8)
** Reading beyond end of file, clamping size 32768 to 2232
00000000=sceAtracDecodeData(0, 08d51084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f424b8], 09faa990[000002e8], 09faa994[001509ec])
0=sceAtracAddStreamData(0, 000002e8)
** Reading beyond end of file, clamping size 32768 to 1488
** FF: Frame data doesn't match channel configuration!
** avcodec_decode_audio4: Error decoding audio -1094995529 / bebbb1b7
80630024=sceAtracDecodeData(0, 08d4d084, 09faa980[00000000], 09faa984[00000001], 09faa988[0])
0=sceAtracReleaseAtracID(0)
```

Based on the truncated warnings, we can assume there are one or two more packets left in the file - where a packet is 0x02e8 bytes - that haven't been added yet.  But there are a few strange things.

```
0=sceAtracSetDataAndGetID(08f62db0, 00001000): atrac3+ mono audio
00000000=sceAtracDecodeData(0, 08d4d084, 09faa980[00000800], 09faa984[00000000], 09faa988[4])
0=sceAtracGetStreamDataInfo(0, 09faa98c[08f421d0], 09faa990[000002e8], 09faa994[0014e9f4])
```

So what happens here is that it takes an *existing* id, and sets brand new data on it.  Then when it asks "what more data do you need?", we reply with "data from 0014e9f4 of course".  There's no way we have read up that far yet.

So I guess what's definitely happening here is that setting new data isn't clearing some important state.

-[Unknown]

**Comment by unknownbrackets** [comment]
Actually, no.  That doesn't make sense.

```
53:39:169 Simple-audio N[ME]: HLE\sceAtrac.cpp:1098 0=sceAtracAddStreamData(0, 000002e8)
53:39:170 Simple-audio I[ME]: HLE\sceAtrac.cpp:1861 0=sceAtracSetDataAndGetID(08f62db0, 00001000): atrac3+ stereo audio
```

That shouldn't happen.  Stream 0 is in use and sceAtracSetDataAndGetID always gets a new one.

It grabs that one here:

```c++
0=sceAtracSetDataAndGetID(08f41c00, 00001000): atrac3+ stereo audio
```

I don't see how it could be getting that id back twice...

-[Unknown]

**Comment by unknownbrackets** [comment]
Darn, it's just that I broke something in logging... so that one is probably just id=1.  Oops.

It's probably this one initially then after all:

```
0=sceAtracSetDataAndGetID(08f41c00, 00001000): atrac3+ stereo audio
```

This is making a lot more sense now.  It starts with 3 remaining frames, but then adds 0x5a0 bytes, and has 4 from then on.  It still reads a packet for every one it decodes.

What happens if you change this line:

```c++
atrac->bufferPos_ = atrac->dataOff_ + atrac->bytesPerFrame_;
```

To:

```c++
atrac->bufferPos_ = atrac->dataOff_ + atrac->bytesPerFrame_ * 2;
```

-[Unknown]

**Comment by daniel229** [comment]
Still audio decode error.

[ppsspplog.txt](https://github.com/hrydgard/ppsspp/files/93862/ppsspplog.txt)

**Comment by unknownbrackets** [comment]
Okay, it works if you make it `* 3` then right?

Maybe the issue is that we don't return `80630023`.  What's the log if you change the `bufferPos_` line back to the original, but then change this:

```c++
					if (res == ATDECODE_FAILED) {
						*SamplesNum = 0;
						*finish = 1;
						return ATRAC_ERROR_ALL_DATA_DECODED;
					}
```

To:

```c++
					if (res == ATDECODE_FAILED) {
						*SamplesNum = 0;
						*finish = 0;
						atrac->failedDecode_ = false;
						return ATRAC_ERROR_BUFFER_IS_EMPTY;
					}
```

Which might make it try to enqueue more...

-[Unknown]

**Comment by daniel229** [comment]
<code>* 3</code> still not work.That change also not work.

**Comment by Levan7** [comment]
still an issue on latest dev build 
[InfoLog.zip](https://github.com/hrydgard/ppsspp/files/3667639/InfoLog.zip)

**Comment by Darkwolf1515** [comment]
For me this will freeze when a droid tries to open something on hoth, happens on the latest stable.

**Comment by sum2012** [comment]
Anyone can test this build ? Thanks
This one improve looping in atrac in streaming

This is windows 64 bit test build 2
https://drive.google.com/file/d/1kKVEwousMWlkHlMbN2X3B13tOxsR4VbZ/view?usp=sharing

**Comment by edubois99** [comment]
Sorry if this is necroposting but I just wanted to say I did give the test build Sum2012 linked a try. First attempt the game froze in the usual spot (leaving the Star Destroyer in the Death Star mission) Second time though it worked. Sadly I wasn't running the proper logging or I'd attach that. I'll run some more tests later.

**Comment by sky252536** [comment]
Hola, soy sky, tengo el mismo problema, no se mucho de programacion, me podrian ayudar a solucionar esto?
my mail is saquiastua@gmail.com

**Comment by hrydgard** [comment]
There's a chance that this is fixed since #20138 (though really by #20098). 

Can someone test with the latest build? 1.18.1-1433 or later.

**Comment by ppmeis** [comment]
Tested in latest build. After loading savestate and go into x-wing, game stops (no PPSSPP crash).

Debug log: 

[mylogfile.zip](https://github.com/user-attachments/files/19426744/mylogfile.zip)

**Comment by hrydgard** [comment]
If you loaded an old savestate, you got the old atrac implementation, as we can't switch while the game is running.

So, start up the game again, and load from a real save to test this properly.

**Comment by ppmeis** [comment]
Yeah I just read this in other issue. I download a 100% save and playing the mission.

**Comment by hrydgard** [comment]
So it's working? or not?

**Comment by ppmeis** [comment]
It's working!
